beets/test/fixtures.py

```diff
@@ -0,0 +1,92 @@
+# This file is part of beets.
+# Copyright 2024, Arav K.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""
+This module provides `pytest`-based fixtures for testing Beets.
+"""
+
+from pathlib import Path
+from typing import Iterator
+
+import pytest
+from confuse import Configuration
+
+import beets
+from beets.library import Library
+
+
+@pytest.fixture(scope="session")
+def resource_dir() -> Path:
+    """
+    A fixture returning the resource directory for tests.
+    """
+
+    return Path(__file__).parents[2] / "test" / "rsrc"
+
+
+@pytest.fixture(scope="function")
+def config(
+    tmp_path: Path,
+    monkeypatch: pytest.MonkeyPatch,
+) -> Iterator[Configuration]:
+    """
+    A fixture for Beets configuration.
+    """
+
+    # 'confuse' looks at `HOME`, so we set it to a tmpdir.
+    monkeypatch.setenv("HOME", str(tmp_path))
+
+    # Initialize configuration.
+    config = beets.config
+    config.sources = []
+    config.read(user=False, defaults=True)
+
+    # Set configuration defaults.
+    config["plugins"] = []
+    config["verbose"] = 1
+    config["ui"]["color"] = False
+    config["threaded"] = False
+    config["directory"] = str(tmp_path)
+
+    # Provide the configuration to the test.
+    yield config
+
+    # Reset the global configuration state.
+    beets.config.clear()
+    beets.config._materialized = False
+
+
+@pytest.fixture(scope="function")
+def library(
+    tmp_path: Path,
+    config: Configuration,
+    monkeypatch: pytest.MonkeyPatch,
+) -> Iterator[Library]:
+    """
+    A fixture for the Beets library.
+    """
+
+    # Beets needs a location to store library contents.
+    lib_dir = tmp_path / "lib_dir"
+    lib_dir.mkdir(exist_ok=False)
+    monkeypatch.setenv("BEETSDIR", str(lib_dir))
+
+    # Create the Beets library object.
+    lib = Library(":memory:", str(lib_dir))
```

Comment:
Maybe it'd be worth defining `lib_dir` as a fixture to remove the need to cast to `str` twice?

```python
@pytest.fixture(scope="session")
def lib_dir(tmp_path_factory: pytest.TempPathFactory) -> str:
    return str(tmp_path_factory.mktemp("lib_dir"))
```

Reply:
I didn't want to do this because these interfaces should be switched over to `pathlib` relatively soon: I want to explicitly leave the `str` casts in so that we can notice and replace them later.

Reply:
I thought that we're using pytest's `tmp_path_factory` here instead?

Reply:
> I didn't want to do this because these interfaces should be switched over to `pathlib` relatively soon: I want to explicitly leave the `str` casts in so that we can notice and replace them later.

I agree that a separate fixture may be an overkill here - consider simply defining a variable `lib_dir = tmp_path_factory.mktemp("lib_dir")` to avoid the double cast and make sure that we only need to a single replacement once we switch to `pathlib`.

Reply:
Unfortunately, `monkeypatch.setenv` [requires](https://docs.pytest.org/en/stable/_modules/_pytest/monkeypatch.html#MonkeyPatch.setenv) a `str`.  That will be true even once we switch over to `pathlib`, so at least one `str` cast will always exist.  What we currently have -- where a second cast is performed for `Library` -- is optimal, since it's the only line that will change.